osis-core/src/main/java/com/scality/osis/service/impl/AsyncScalityOsisService.java

```diff
@@ -156,36 +152,30 @@ private void createAttachAdminPolicy(String tenantId, String assumeRoleName,
-                                        URLDecoder.decode(DEFAULT_ADMIN_POLICY_DOCUMENT))) {
-                                        logger.debug("Default Admin Policy Document changed, before: {}, current: {}",
-                                                getAdminPolicyVersionResult.getPolicyVersion().getDocument(),
-                                                DEFAULT_ADMIN_POLICY_DOCUMENT);
-                                        CreatePolicyVersionRequest createAdminPolicyVersionRequest = ScalityModelConverter.toCreateAdminPolicyVersionRequest(tenantId);
-                                        logger.debug("[Vault] Create new Admin Policy Version Request:{}", new Gson().toJson(createAdminPolicyVersionRequest));
-                                        CreatePolicyVersionResult createAdminPolicyVersionResult = iamClient.createPolicyVersion(createAdminPolicyVersionRequest);
-                                        logger.debug("[Vault] Create new Admin Policy Version Result:{}", new Gson().toJson(createAdminPolicyVersionResult));
-                                        return;
-                                }
+                                logger.debug("[Vault] Admin Policy already exists, detach role policy, delete it and create a new one");
+
+                                DetachRolePolicyRequest detachRolePolicyRequest = ScalityModelConverter.toDetachAdminPolicyRequest(
+                                        ScalityModelConverter.toAdminPolicyArn(tenantId),
+                                        assumeRoleName);
+                                logger.debug("[Vault] Detach Admin Policy Request:{}", new Gson().toJson(detachRolePolicyRequest));
+
+                                DetachRolePolicyResult detachRolePolicyResult = iamClient.detachRolePolicy(detachRolePolicyRequest);
+                                logger.debug("[Vault] Detach Admin Policy response:{}", new Gson().toJson(detachRolePolicyResult));
+
+                                DeletePolicyRequest deletePolicyRequest = ScalityModelConverter.toDeleteAdminPolicyRequest(tenantId);
+                                logger.debug("[Vault] Delete Admin Policy Request:{}", new Gson().toJson(detachRolePolicyRequest));
+
+                                DeletePolicyResult deletePolicyResult = iamClient.deletePolicy(deletePolicyRequest);
+                                logger.debug("[Vault] Delete Admin Policy Response:{}", new Gson().toJson(deletePolicyResult));
+
+                                adminPolicyResult = iamClient.createPolicy(createAdminPolicyRequest);
+                                logger.debug("[Vault] Create Admin Policy response:{}", new Gson().toJson(adminPolicyResult));
+
+                        } else {
+                                logger.error("Cannot create admin policy for Tenant ID:{}. " +
+                                        "Exception details:{}", tenantId, e.getMessage());
```

Comment:
Is this backward compatible with the old way of policy attached?

Reply:
yes it is. and also I am thinking sometime later we can use a service account for osis, seed the osis needed role and policy into account at account creation time. this way we don't need to maintain policy at osis side